core/src/main/java/com/google/cloud/sql/ApiClientFactory.java

```diff
@@ -0,0 +1,26 @@
+/*
+ * Copyright 2023 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.sql;
+
+import com.google.api.client.googleapis.services.json.AbstractGoogleJsonClient;
+import com.google.api.client.http.HttpRequestInitializer;
+import com.google.api.services.sqladmin.SQLAdmin;
+import com.google.common.base.Optional;
+
+public interface ApiClientFactory {
+  AbstractGoogleJsonClient create(HttpRequestInitializer credentials);
```

Comment:
Let's make this return an `SQLAdmin` type -- the abstract client isn't what we'll be working with.

Reply:
It's an abstract client so that we can have a stub factory that returns a stub that implements the abstract client.

Reply:
We can pass a [MockHttpTransport](https://javadoc.io/static/com.google.http-client/google-http-client/1.43.0/com/google/api/client/testing/http/MockHttpTransport.html) into the `SqlAdmin` class to create the equivalent of a stub.

Reply:
And see [here](https://developers.google.com/resources/api-libraries/documentation/sqladmin/v1beta4/java/latest/com/google/api/services/sqladmin/SQLAdmin.Builder.html#Builder-com.google.api.client.http.HttpTransport-com.google.api.client.json.JsonFactory-com.google.api.client.http.HttpRequestInitializer-) for the SqlAdmin builder that takes an HttpTransport.